### DISCUSSION TITLE
Prevent copying to .git/lfs/objects after git add

### DISCUSSION THREAD

**Discussion by Lenocc** [git-lfs/git-lfs#5576] [Ideas] [answered]
One significant issue with git lfs is the need for additional disk space, specifically after executing git add, where lfs files are copied to .git/lfs/objects/..., effectively doubling the data size on the local disk. It may lead to an inability to commit changes due to insufficient space for copies in .git/lfs/. A common sequence of actions when dealing with large files includes:
`git add .`
`git commit "xyz"`
`git push`
`git lfs prune --force`

When working with files of 10GB, an additional 10GB of free space on the disk is needed for the copy operation. Shortly afterward, changes are sent to the server, and the copies created a moment ago are deleted. This is not an optimal or disk-friendly solution and does not contribute to prolonging disk lifespan. Does git-lfs plan to address this issue in the future and is it even possible?

git lfs push -a <lfs files> -m "text" -f

**Comment by bk2204** [comment, marked as answer]
There's really no way to avoid this because `git add` filters the file through the clean filter, at which point Git LFS gets ahold of it and has to store it.  Git doesn't offer us the ability to do copy-on-write here, and only some file systems even offer copy-on-write (notably, APFS, Btrfs, and ReFS, but not NTFS or ext4).

You can use `git lfs dedup`, which replaces what's in the working tree with a copy-on-write snapshot of what's in the datastore if that's supported on your system, but that will necessarily require that Git re-read and re-hash the objects (which will be cheaper on systems supporting SHA-NI in Git LFS 3.5).  It will, however, avoid the duplication provided the file in the working tree is not modified.

**Reply by Lenocc** [reply]
Thank you for the explanation, that's a pity

**Reply by bk2204** [reply]
I've opened #5578 for us to look at making some changes to Git to help copy-on-write functionality be able to be implemented in Git LFS.  We have no firm timeline for implementing this, but it should make the situation better once it is implemented.
